### PR TITLE
[Merged by Bors] - feat: port Order.Filter.Lift

### PR DESCRIPTION
**Pull request by urkud**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/Filter/Lift.lean [hunk 1]
```diff
@@ -438,30 +427,19 @@ section Prod
...
+
 theorem prod_same_eq : f ×ᶠ f = f.lift' fun t : Set α => t ×ˢ t :=
-  prod_def.trans <|
-    lift_lift'_same_eq_lift' (fun s => monotone_const.set_prod monotone_id) fun t =>
-      monotone_id.set_prod monotone_const
+  f.basis_sets.prod_self.eq_binfᵢ
 #align filter.prod_same_eq Filter.prod_same_eq
 
-/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
-theorem mem_prod_same_iff {s : Set (α × α)} : s ∈ f ×ᶠ f ↔ ∃ t ∈ f, t ×ˢ t ⊆ s := by
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Lift.lean]: What happened to this lemma?

File: Mathlib/Order/Filter/Lift.lean [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+
+-- porting note: use `∃ i, p i ∧ _` instead of `∃ i (hi : p i), _`
+/-- If `(p : ι → Prop, s : ι → set α)` is a basis of a filter `f`, `g` is a monotone function
+`set α → filter γ`, and for each `i`, `(pg : β i → Prop, sg : β i → set α)` is a basis
+of the filter `g (s i)`, then `(λ (i : ι) (x : β i), p i ∧ pg i x, λ (i : ι) (x : β i), sg i x)`
+is a basis of the filter `f.lift g`.
+
+This basis is parametrized by `i : ι` and `x : β i`, so in order to formulate this fact using
+`has_basis` one has to use `Σ i, β i` as the index type, see `filter.has_basis.lift`.
+This lemma states the corresponding `mem_iff` statement without using a sigma type. -/
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Lift.lean:46]:
```suggestion
/-- If `(p : ι → Prop, s : ι → Set α)` is a basis of a filter `f`, `g` is a monotone function
`Set α → Filter γ`, and for each `i`, `(pg : β i → Prop, sg : β i → Set α)` is a basis
of the filter `g (s i)`, then `(λ (i : ι) (x : β i), p i ∧ pg i x, λ (i : ι) (x : β i), sg i x)`
is a basis of the filter `f.lift g`.

This basis is parametrized by `i : ι` and `x : β i`, so in order to formulate this fact using
`HasBasis` one has to use `Σ i, β i` as the index type, see `Filter.HasBasis.Lift`.
This lemma states the corresponding `mem_iff` statement without using a sigma type. -/
```

File: Mathlib/Order/Filter/Lift.lean [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+#align filter.has_basis.mem_lift_iff Filter.HasBasis.mem_lift_iffₓ
+
+/-- If `(p : ι → Prop, s : ι → set α)` is a basis of a filter `f`, `g` is a monotone function
+`set α → filter γ`, and for each `i`, `(pg : β i → Prop, sg : β i → set α)` is a basis
+of the filter `g (s i)`, then `(λ (i : ι) (x : β i), p i ∧ pg i x, λ (i : ι) (x : β i), sg i x)`
+is a basis of the filter `f.lift g`.
+
+This basis is parametrized by `i : ι` and `x : β i`, so in order to formulate this fact using
+`has_basis` one has to use `Σ i, β i` as the index type. See also `filter.has_basis.mem_lift_iff`
+for the corresponding `mem_iff` statement formulated without using a sigma type. -/
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Lift.lean:65]:
```suggestion
/-- If `(p : ι → Prop, s : ι → Set α)` is a basis of a filter `f`, `g` is a monotone function
`Set α → Filter γ`, and for each `i`, `(pg : β i → Prop, sg : β i → Set α)` is a basis
of the filter `g (s i)`, then `(λ (i : ι) (x : β i), p i ∧ pg i x, λ (i : ι) (x : β i), sg i x)`
is a basis of the filter `f.lift g`.

This basis is parametrized by `i : ι` and `x : β i`, so in order to formulate this fact using
`has_basis` one has to use `Σ i, β i` as the index type. See also `Filter.HasBasis.mem_lift_iff`
for the corresponding `mem_iff` statement formulated without using a sigma type. -/
```

File: Mathlib/Order/Filter/Lift.lean [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+theorem tendsto_lift' {m : γ → β} {l : Filter γ} :
+    Tendsto m l (f.lift' h) ↔ ∀ s ∈ f, ∀ᶠ a in l, m a ∈ h s := by
+  simp only [Filter.lift', tendsto_lift, tendsto_principal, comp]
+#align filter.tendsto_lift' Filter.tendsto_lift'
+
+theorem HasBasis.lift' {ι} {p : ι → Prop} {s} (hf : f.HasBasis p s) (hh : Monotone h) :
+    (f.lift' h).HasBasis p (h ∘ s) :=
+  ⟨fun t => (hf.mem_lift_iff (fun i => hasBasis_principal (h (s i)))
+    (monotone_principal.comp hh)).trans <| by simp only [exists_const, true_and, comp]⟩
+
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Lift.lean:265]:
```suggestion
```

File: Mathlib/Order/Filter/Lift.lean [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+theorem lift'_inf_principal_eq {h : Set α → Set β} {s : Set β} :
+    f.lift' h ⊓ 𝓟 s = f.lift' fun t => h t ∩ s := by
+  simp only [Filter.lift', Filter.lift, (· ∘ ·), ← inf_principal, infᵢ_subtype', ← infᵢ_inf]
+#align filter.lift'_inf_principal_eq Filter.lift'_inf_principal_eq
+
+theorem lift'_neBot_iff (hh : Monotone h) : NeBot (f.lift' h) ↔ ∀ s ∈ f, (h s).Nonempty :=
+  calc
+    NeBot (f.lift' h) ↔ ∀ s ∈ f, NeBot (𝓟 (h s)) := lift_neBot_iff (monotone_principal.comp hh)
+    _ ↔ ∀ s ∈ f, (h s).Nonempty := by simp only [principal_neBot_iff]
+
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Lift.lean:388]:
```suggestion
```

File: Mathlib/Order/Filter/Lift.lean [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+theorem prod_lift'_lift' {f₁ : Filter α₁} {f₂ : Filter α₂} {g₁ : Set α₁ → Set β₁}
+    {g₂ : Set α₂ → Set β₂} (hg₁ : Monotone g₁) (hg₂ : Monotone g₂) :
+    f₁.lift' g₁ ×ᶠ f₂.lift' g₂ = f₁.lift fun s => f₂.lift' fun t => g₁ s ×ˢ g₂ t :=
+  calc
+    f₁.lift' g₁ ×ᶠ f₂.lift' g₂ = f₁.lift fun s => f₂.lift fun t => 𝓟 (g₁ s) ×ᶠ 𝓟 (g₂ t) :=
+      prod_lift_lift (monotone_principal.comp hg₁) (monotone_principal.comp hg₂)
+    _ = f₁.lift fun s => f₂.lift fun t => 𝓟 (g₁ s ×ˢ g₂ t) := by
+      { simp only [prod_principal_principal] }
+
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Lift.lean:468]:
```suggestion
```

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

File: Mathlib/Order/Filter/Lift.lean (hunk 1 above)
Comment (urkud) [reply Mathlib/Order/Filter/Lift.lean]: Deduplicated with `mem_prod_self_iff`.

**Pull request closed without merging**